standards/ICRC-3/README.md

```diff
@@ -0,0 +1,206 @@
+independently without breaking them, it also means that a client may not receive all the information that the server is sending, e.g. in case the client schema lacks some fields that the server schema has.
+
+This loss of information is not an option for `ICRC-3`. The client must receive the same exact data the server sent. For this reason, `ICRC-3` introduces the `Value` type which never changes: 
+
+```
+type Value = variant { 
+    Blob : blob; 
+    Text : text; 
+    Nat : nat; // do we need this or can we just use Int?
+    Int : int;
+    Array : vec Value; 
+    Map : vec record { text; Value }; 
+};
+```
+
+Servers must serve the transaction log as list of `Value` where each `Value` represent a single transaction in the transaction log.
+
+## Value Hash
+
+`ICRC-3` specifies a standard hash function over `Value`.
+
+This hash function should be used by Ledgers to calculate the hash of the parent of a transaction and by clients to verify the downloaded transaction log.
+
+The hash function is the [representation-independent hashing of structured data](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) used by the IC:
+- the hash of a `Blob` is the hash of the bytes themselves
+- the hash of a `Text` is the hash of the bytes representing the text
+- the hash of a `Nat` is the [`leb128`](https://en.wikipedia.org/wiki/LEB128#Unsigned_LEB128) encoding of the number
+- the hash of an `Int` is the [`sleb128`](https://en.wikipedia.org/wiki/LEB128#Signed_LEB128) encoding of the number
+- the hash of an `Array` is the hash of the concatenation of the hashes of all the elements of the array
+- the hash of a `Map` is the hash of the concatenation of all the hashed items of the map sorted. A hashed item is the tuple composed by the hash of the key and the hash of the value.
+
+## Interaction with other standards
+
+Each standard that adheres to `ICRC-3` must define the list of transactions types they define and/or extend together with their `Value` schema and the function that converts a [`Value`](#value) to that type. Transaction types are well-typed records that are easy to consume by clients.
+
+`Value`s representing transactions must have the following schema properties:
+1. they must be of type `Map`
+1. they must have a field `tx: Map` which describes the transaction
+1. they must have a field `op: Text` inside `tx` which describes the type of transactions, e.g. "ICRC1_Burn" for `ICRC1_Burn` transactions
+1. all transactions with `index > 0` must have a top-level field called `phash: Blob` which contains the [hash](#value-hash) of the previous transactions. 
+
+In other words the schema must be:
+
+```
+type ICRC1_Transaction = {
+    "phash" : Blob?;
+    "tx" : {
+        "op" : Text;
+        ...
+    };
+};
```

Comment:
Because it's not a candid type, it's a schema of a Value object.

The `Value` type is like JSON, the schema here is like the JSON schema definition. It describes the shape of the value, which could not be expressed statically in Candid, so we had to invent a separate meta-language.

Reply:
> Because it's not a candid type, it's a schema of a Value object.

That's exactly what I'm challenging.
OK, you need the `Value` type because you need to be able to express future inclusions.
But that doesn't mean that everything immediately needs to use the `Value` type.
Things that you already know must exist can be included in a proper type and the `Value` type can be used for future metadata.
Using this untyped type just makes things hard, so I'm suggesting to only use it where you have.

Reply:
> untyped type

IMO, mixing "typed types" and "untyped types" will only create more confusion because validators will have to convert the data into a uniform "untyped" representation anyway to compute the block hash.
Providing a library for decoding typed records from Values is a simpler and cleaner design choice.

Reply:
I don't think the "untyped type" is harder to handle than the mixed one in this case. The problem with the mixed type is how to compute the block hash. I see two ways:
1. convert the mixed type to a "untyped type" but then you are basically writing the same code you need to go from "untyped type" to "typed type" just the other way around
2. Provide an algorithm that works on both "untyped type" and "typed type" and their mix (e.g. given something that resemble a structure, create a list of all fields hash, sort them and return the hash.). This would be as hard as the other two options if not harder because it's weirder. In your example, the metadata field should be "transparent" and its content should be treated as part of the "typed type"